AbortController/Signal Callback Fix (#22)
BabylonJS/JsRuntimeHost: Documentation/DebugJavascriptDevTools.md

Reviewer: nit: I know you copied this from Babylon Native, but can you rename the file to `DebugJavaScriptDevTools.md` with capital S? Also in other places in the readme, please use JavaScript and not Javascript.

# JsRuntimeHost Script Debugging
You can use Chromium DevTools to debug your JavaScript while running with the V8 JavaScript engine (Default Engine for Android).

## Add the Remote Target URL
On whichever Chromium-based browser you're using (Edge or Chrome), navigate to `about://inspect`.

> Click `Configure...` and in the target discovery settings add an entry `localhost:{port}`, where `{port}` is consistent with the one supplied to the call in [AppRuntime_V8.cpp](https://github.com/BabylonJS/JsRuntimeHost/blob/f487c7b3f89b407e95a53543a06a34f1a1fbb860/Core/AppRuntime/Source/AppRuntime_V8.cpp#L85) and [AppRuntime_JSI.cpp](https://github.com/BabylonJS/JsRuntimeHost/blob/09c5acc49a6a64dec0e62075aebac6ad0a32aa93/Core/AppRuntime/Source/AppRuntime_JSI.cpp#L39).

Reviewer: The JSI one isn't hooked up yet, so maybe don't put that here yet. Maybe just put a hard coded 5643 for the port for now. If we decide that this port needs to be configurable, we can update the code and docs.

![Target discovery settings](Images/DevTools/chrome-targets.png)

## Setting Up To Debug
Make sure to change the line in [tests.js](https://github.com/BabylonJS/JsRuntimeHost/blob/f487c7b3f89b407e95a53543a06a34f1a1fbb860/Tests/UnitTests/Scripts/tests.js#L2) to true, or else the DevTools won't have enough time to attach to the Javascript Instance. If you are debugging your own script, you can also use [setTimeout()](https://developer.mozilla.org/en-US/docs/web/api/settimeout) to delay the execution of the code you are trying to debug while you wait to attach the debugger.

## Debugging on Android/Emulator

> Download and unzip [ADB](https://developer.android.com/tools/releases/platform-tools) and run the following commands in that folder while your android device is plugged in (or emulator is on in Android Studio)

Reviewer: nit:
```suggestion
Install [ADB](https://developer.android.com/tools/releases/platform-tools) and run the following command with an Android device plugged in or Android emulator running.
```

```
adb forward tcp:5643 tcp:5643
```

> Note, every time you unplug and re-plug your device these ADB commands need to be run again.

Reviewer: ```suggestion
Note, every time the Android device is unpluggeg and re-plugged, this ADB command need to be run again.
```

## Click Inspect
Run the UnitTests application and wait for the DevTools page to recognize the inspector server spun up by the app. Once you see the app pop up in the list of Remote Targets, you may click inspect to open up a DevTools window.

![Ready to inspect](Images/DevTools/chrome-inspect.png)

## DevTool Settings

> There is also a setting that needs to be enabled in the DevTools. Go into the settings by clicking on the gear in the top right corner and check this box in the preferences tab.

Reviewer: ```suggestion
There is also a setting that needs to be enabled in the DevTools. Go into the settings by clicking on the gear in the top right corner and check this box in the preferences tab. This only needs to be done once.
```

# JsRuntimeHost Script Debugging
You can use Chromium DevTools to debug your JavaScript while running with the V8 JavaScript engine (Default Engine for Android).

## Add the Remote Target URL

> Depending on which Chromium-based browser you're using (Edge or Chrome), navigate to `edge://inspect` or `chrome://inspect`.

Reviewer: I know you copied this but a cross-browser way to do this is to use `about://inspect`.

![Target discovery settings](Images/DevTools/chrome-targets.png)

## Setting Up To Debug
Make sure to change the line in [tests.js](https://github.com/BabylonJS/JsRuntimeHost/blob/f487c7b3f89b407e95a53543a06a34f1a1fbb860/Tests/UnitTests/Scripts/tests.js#L2) to true, or else the DevTools won't have enough time to attach to the Javascript Instance. If you are debugging your own script, you can also use [setTimeout()](https://developer.mozilla.org/en-US/docs/web/api/settimeout) to delay the execution of the code you are trying to debug while you wait to attach the debugger.

Also, currently a bug, but line 574 in file [V8InspectorAgent.cpp](https://github.com/BabylonJS/JsRuntimeHost/blob/f487c7b3f89b407e95a53543a06a34f1a1fbb860/Core/AppRuntime/V8Inspector/Source/V8InspectorAgent.cpp#L574) needs to be commented out if it hasn't been already.
```

> CHECK_EQ(State::kAccepting, state_);

Reviewer: I already removed this in the code.

```

## Click Inspect
Run the UnitTests application and wait for the DevTools page to recognize the inspector server spun up by the app. Once you see the app pop up in the list of Remote Targets, you may click inspect to open up a DevTools window.

![Ready to inspect](Images/DevTools/chrome-inspect.png)

## View Logs, Set Breakpoints, Profile, and More
Once you have the DevTools window open, you can use it in the same way as if you were inspecting a web page.
For more information, see this documentation from Google on [how to debug JavaScript using Chrome DevTools](https://developer.chrome.com/docs/devtools/javascript/).

![DevTools window](Images/DevTools/chrome-debugger.png)

## DevTool Settings

If you have issues loading some of your Javascript files double check that this setting is enabled. Go into the settings by clicking on the gear in the top right corner and check this box in the preferences tab.

> ![DevTools window](Images/DevTools/devtools-settings.png)

Reviewer: It sounds like this is optional, but it really isn't (at least not for default usage). Maybe it would be good to include as one of a main step right after clicking inspect.

## Debugging on Android

> Download and unzip ADB and run the following commands in that folder while your android device is plugged in (or emulator is on in Android Studio)

Reviewer 2: Not a big deal, but might be useful to have a link to ADB.

```
adb forward tcp:5643 tcp:5643

> adb reverse tcp:8000 tcp:8000

Reviewer: This one isn't necessary. Also, if `adb forward` isn't run, inspect won't show up. I think this instruction should be earlier.
